File: python/bag_of_tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: list[int], power: int) -> int:
        """time complexity: O(n)"""
        tokens.sort()
        result: int = 0
        score: int = 0
        l: int = 0
        r: int = len(tokens) - 1
        while l <= r:
            if power >= tokens[l]:
                power -= tokens[l]
                l += 1
                score += 1
                result = max(result, score)
            elif score > 0:
                power += tokens[r]
                score -= 1
                r -= 1
            else:
                break
        return result


def test_bagOfTokensScore_case_1():
    # arrange
    tokens: list[int] = [100]
    power: int = 50
    expected: int = 0

    # act
    solution = Solution()
    actual: int = solution.bagOfTokensScore(tokens, power)

    # assert
    assert expected == actual


def test_bagOfTokensScore_case_2():
    # arrange
    tokens: list[int] = [200, 100]
    power: int = 150
    expected: int = 1

    # act
    solution = Solution()
    actual: int = solution.bagOfTokensScore(tokens, power)

    # assert
    assert expected == actual


def test_bagOfTokensScore_case_3():
    # arrange
    tokens: list[int] = [100, 200, 300, 400]
    power: int = 200
    expected: int = 2

    # act
    solution = Solution()
    actual: int = solution.bagOfTokensScore(tokens, power)

    # assert
    assert expected == actual


def test_bagOfTokensScore_case_4():
    # arrange
    tokens: list[int] = []
    power: int = 85
    expected: int = 0

    # act
    solution = Solution()
    actual: int = solution.bagOfTokensScore(tokens, power)

    # assert
    assert expected == actual


def test_bagOfTokensScore_case_4():
    # arrange
    tokens: list[int] = []
    power: int = 85
    expected: int = 0

    # act
    solution = Solution()
    actual: int = solution.bagOfTokensScore(tokens, power)

    # assert
    assert expected == actual

File: rust/src/code/bag_of_tokens.rs
struct Solution;

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn bag_of_tokens_score(mut tokens: Vec<i32>, mut power: i32) -> i32 {
        use std::cmp::max;
        if tokens.len() == 0 {
            return 0;
        }
        tokens.sort();
        let mut result: i32 = 0;
        let mut score: i32 = 0;
        let mut l: usize = 0;
        let mut r: usize = tokens.len() - 1;
        while l <= r {
            if power >= tokens[l] {
                power -= tokens[l];
                score += 1;
                l += 1;
                result = max(result, score);
            } else if score > 0 {
                power += tokens[r];
                score -= 1;
                r -= 1;
            } else {
                break;
            }
            dd
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bag_of_tokens_score_case_1() {
        // arrange
        let tokens: Vec<i32> = vec![100];
        let power: i32 = 50;
        let expected: i32 = 0;

        // act
        let actual: i32 = Solution::bag_of_tokens_score(tokens, power);

        // arrange
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_bag_of_tokens_score_case_2() {
        // arrange
        let tokens: Vec<i32> = vec![200, 100];
        let power: i32 = 150;
        let expected: i32 = 1;

        // act
        let actual: i32 = Solution::bag_of_tokens_score(tokens, power);

        // arrange
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_bag_of_tokens_score_case_3() {
        // arrange
        let tokens: Vec<i32> = vec![100, 200, 300, 400];
        let power: i32 = 200;
        let expected: i32 = 2;

        // act
        let actual: i32 = Solution::bag_of_tokens_score(tokens, power);

        // arrange
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_bag_of_tokens_score_case_4() {
        // arrange
        let tokens: Vec<i32> = vec![];
        let power: i32 = 85;
        let expected: i32 = 0;

        // act
        let actual: i32 = Solution::bag_of_tokens_score(tokens, power);

        // arrange
        assert_eq!(expected, actual);
    }
}
